expression/generator/compare_vec.go

```diff
@@ -0,0 +1,316 @@
+		case types.CompareFloat64(arg0[i], arg1[i]) == 0:{{ else if eq .type.ETName "String" }}
+		case types.CompareString(buf0.GetString(i), buf1.GetString(i)) == 0:{{ else if eq .type.ETName "Duration" }}
+		case types.CompareDuration(arg0[i], arg1[i]) == 0:{{ else if eq .type.ETName "Datetime" }}
+		case arg0[i].Compare(arg1[i]) == 0:{{ else if eq .type.ETName "Decimal" }}
+		case arg0[i].Compare(&arg1[i]) == 0:{{ end }}
+			i64s[i] = 1
+		}
+	}
+	return nil
+}
+
+func (b *builtin{{ .compare.CompareName }}{{ .type.TypeName }}Sig) vectorized() bool {
+	return true
+}
+`))
+
+const builtinCompareVecTestHeader = `import (
+	"testing"
+
+	. "github.com/pingcap/check"
+	"github.com/pingcap/parser/ast"
+	"github.com/pingcap/tidb/types"
+)
+
+var vecGeneratedBuiltinCompareCases = map[string][]vecExprBenchCase{
+`
+
+var builtinCompareVecTestFuncHeader = template.Must(template.New("").Parse(`	ast.{{ .CompareName }}: {
+`))
+
+var builtinCompareVecTestCase = template.Must(template.New("").Parse(`		{retEvalType: types.ETInt, childrenTypes: []types.EvalType{types.ET{{ .ETName }}, types.ET{{ .ETName }}}},
+`))
+
+var builtinCompareVecTestFuncTail = `	},
+`
+
+var builtinCompareVecTestTail = `}
+
+func (s *testEvaluatorSuite) TestVectorizedGeneratedBuiltinCompareEvalOneVec(c *C) {
+	testVectorizedEvalOneVec(c, vecGeneratedBuiltinCompareCases)
+}
+
+func (s *testEvaluatorSuite) TestVectorizedGeneratedBuiltinCompareFunc(c *C) {
+	testVectorizedBuiltinFunc(c, vecGeneratedBuiltinCompareCases)
+}
+
+func BenchmarkVectorizedGeneratedBuiltinCompareEvalOneVec(b *testing.B) {
+	benchmarkVectorizedEvalOneVec(b, vecGeneratedBuiltinCompareCases)
+}
+
+func BenchmarkVectorizedGeneratedBuiltinCompareFunc(b *testing.B) {
+	benchmarkVectorizedBuiltinFunc(b, vecGeneratedBuiltinCompareCases)
+}
+`
+
+type CompareContext struct {
+	// Describe the name of CompareContext(LT/LE/GT/GE/EQ/NE/NullEQ)
+	CompareName string
+	// Compare Operators
+	Operator string
+}
+
+var comparesMap = []CompareContext{
+	{CompareName: "LT", Operator: "<"},
+	{CompareName: "LE", Operator: "<="},
+	{CompareName: "GT", Operator: ">"},
+	{CompareName: "GE", Operator: ">="},
+	{CompareName: "EQ", Operator: "=="},
+	{CompareName: "NE", Operator: "!="},
+	{CompareName: "NullEQ"},
+}
+
+var typesMap = []TypeContext{
+	TypeReal,
+	TypeDecimal,
+	TypeString,
+	TypeDatetime,
+	TypeDuration,
+	TypeJSON,
+}
+
+func generateDotGo(fileName string, compares []CompareContext, types []TypeContext) (err error) {
+	w := new(bytes.Buffer)
+	w.WriteString(header)
+	w.WriteString(newLine)
+	w.WriteString(builtinCompareImports)
+
+	var ctx = make(map[string]interface{})
+	for _, compareCtx := range compares {
+		for _, typeCtx := range types {
+			ctx["compare"] = compareCtx
+			ctx["type"] = typeCtx
+			if compareCtx.CompareName == "NullEQ" {
+				if typeCtx.ETName == "Real" ||
+					typeCtx.ETName == "Decimal" ||
+					typeCtx.ETName == "String" {
+					continue
+				}
```

Comment:
> It's better to unify them into your generator.

We can delete these implementations in `builtin_compare_vec.go` and unify them into your generator. So each function of `NullEQ` will be maintained by `go generate` in the future.